core/storage/couchStorageHelpers.go

```diff
@@ -0,0 +1,582 @@
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return notFound
+		}
+		return err
+	}
+
+	for rows.Next() {
+
+		var tempObject couchDestinationObject
+		err = rows.ScanDoc(&tempObject)
+		if err != nil {
+			return err
+		}
+
+		if _, err := db.Delete(context.TODO(), tempObject.ID, tempObject.Rev); err != nil {
+			return err
+		}
+	}
+
+	if err := rows.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) deleteAllNotificationObjects(query interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return notFound
+		}
+		return err
+	}
+
+	for rows.Next() {
+
+		var tempObject couchNotificationObject
+		err = rows.ScanDoc(&tempObject)
+		if err != nil {
+			return err
+		}
+
+		if _, err := db.Delete(context.TODO(), tempObject.ID, tempObject.Rev); err != nil {
+			return err
+		}
+	}
+
+	if err := rows.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) getAttachment(id string) (*kivik.Attachment, common.SyncServiceError) {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+	attachment, err := db.GetAttachment(context.TODO(), id, id)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return nil, notFound
+		}
+		return nil, err
+	}
+	return attachment, nil
+}
+
+func (store *CouchStorage) retrievePolicies(query interface{}) ([]common.ObjectDestinationPolicy, common.SyncServiceError) {
+	results := []couchObject{}
+
+	if err := store.findAll(query, &results); err != nil {
+		if err == notFound {
+			return nil, nil
+		}
+		return nil, &Error{fmt.Sprintf("Failed to fetch the objects with a Destination Policy. Error: %s", err)}
+	}
+
+	objects := make([]common.ObjectDestinationPolicy, len(results))
+	for index, oneResult := range results {
+		destinationList := make([]common.DestinationsStatus, len(oneResult.Destinations))
+		for destIndex, destination := range oneResult.Destinations {
+			destinationList[destIndex] = common.DestinationsStatus{
+				DestType: destination.Destination.DestType, DestID: destination.Destination.DestID,
+				Status: destination.Status, Message: destination.Message,
+			}
+		}
+		objects[index] = common.ObjectDestinationPolicy{
+			OrgID: oneResult.MetaData.DestOrgID, ObjectType: oneResult.MetaData.ObjectType, ObjectID: oneResult.MetaData.ObjectID,
+			DestinationPolicy: oneResult.MetaData.DestinationPolicy, Destinations: destinationList,
+		}
+	}
+	return objects, nil
+}
+
+func (store *CouchStorage) addUsersToACLHelper(aclType string, orgID string, key string, users []common.ACLentry) common.SyncServiceError {
+	var id string
+	if key == "" {
+		id = orgID + ":" + aclType + ":*"
+	} else {
+		id = orgID + ":" + aclType + ":" + key
+	}
+
+	if trace.IsLogging(logger.TRACE) {
```

Comment:
There's not much debug tracing in this code, do you think you need more?

Reply:
Here also I followed a pattern similar to the Mongo implementation...what is a good way to find a balance between too much and too little debug tracing?

Reply:
If there is a function that was particularly hard to implement or difficult to get working through unit tests, then it might need more tracing because it is likely to be re-visited in the future due to a bug. Think about it this way; would you rather get awakened at 2AM by your boss because no one else can debug a problem, or would you rather have enough tracing in there such that someone else can debug a problem and you get to keep sleeping.

Reply:
That's an interesting way to think about it! I'll add more debug tracing in the code, thank you